dpg4x/pkg_common/dpg4x.1
.\" .ig
.\" dpg4x manual page.
.\" 
.\" Félix Medrano <xukosky@yahoo.es>, 2010.
.\" 
.\" This next line loads the www macro package so that the .URL macro can be used.
.\" ..
.\" .mso www.tmac
.TH "DPG for X" "1" "12 October 2012"
.SH "NAME"
.
dpg4x \- A GUI to encode videos into DPG format

.
.SH "SYNOPSIS"
.
.B dpg4x
\fI[Options]\fR \fIfile1\fR \fI[file2]\fR \fI[...]\fR (starts GUI if no options given)

.
.SH "DESCRIPTION"
.
Dpg4x is a program that tries to make the DPG video creation as easy as possible.
A DPG video is the video format that you need if you want to play movies on your Nintendo DS.

Dpg4x relies on mplayer/mencoder to do it's work, so any video file playable
with mplayer can be converted to DPG. 

You can also encode the DPG video files to AVI format or replace the thumbnail image
embedded in a DPG4 video using the command line options.
.
.SH "FEATURES"
 \- Simple GUI suitable for beginners.
 \- Includes lots of options for the advanced user.
 \- Command line options to encode videos from batch jobs.
 \- Supports individual per-media settings when needed.
 \- Multiple DPG version support, from DPG0 to DPG4.
 \- Can process any video file playable by mplayer.
 \- Encodes your DVD and VCD directly.
 \- Includes subtitles support.
 \- Ability to preview the encoding settings.
 \- Drag and drop support from your favourite file manager.
 \- Batch processing.
 \- Multiplatform.
 \- Multilanguage.

.SH "USAGE"
.
.TP
.B Media Sources List
.
If you run dpg4x, the first thing that you will see is the MAIN section. You can
switch from one section to another with the icon-buttons at the top.

The MAIN section includes a list, with all the media sources that you want to
encode. To add a media source to the list, use the "Add Media" button. This will
enable the "Add File", "Add DVD" and "Add VCD" buttons. To delete a media source
from the list, select the media and use the "Delete Media" button. 

To add a file, just browse your file system and select it. To add a DVD or a VCD
you need to select the track and the disc reader device. When adding a DVD, you
can filter by a range of chapters.

Need more? Now you can also add files with drag-and-drop from your favourite
file manager!
.TP
.B Start Encoding
.
When you have completed the list, just use the "Start Encoding" button. If you
want to setup the options for the encoding, you can select the DPG Version and
the quality that you want to use. The DPG videos will be stored in the same
folder as the original video files, but you can change this behaviour in the
MISC section. This is a must when you are using DVD or VCD sources.

If the source file already is a DPG file it will be converted to an AVI file
without changing any video or audio streams. Encoding settings do not apply for
DPG files.

.TP
.B Preview Modes
.
With the "Play Media" and "DPG Preview" buttons at the MAIN section you can
play the media sources before encoding.

However, they work in a different way, while on the one hand "Play Media" just
plays the media source "as is" using mplayer, the "DPG Preview" button encodes
the beginning of the video, using the encoding settings you have selected, and
then plays it as a DPG-like video. This is specially useful for testing your
settings and seeing how your video would like. You can setup the length of the
preview from the MISC section.

The "Info" button shows more detailed mplayer information about the video. For DPG
files there is additional header information and a thumbnail view. This dialog also 
includes a function for replacing the thumbnail in a DPG file.

.TP
.B Default Settings
.
When you need to adjust the encoding process to your needs, dpg4x offers two
ways to configure it.

If you want your settings to be applied to all the videos you encode, you just
need to switch to another section and adjust the default settings. Remember that
all the sections are always available via the icon-buttons at the top.

The VIDEO section includes advanced video options. e.g. You can change the video
bitrate, and the FPS. In the AUDIO section you can select the audio track to
encode and adjust the audio quality. The SUBTITLES section includes all the
options you need to add subtitles to the videos, and the MISC section includes
the options that didn't fit in the other sections, like the tmp folder that
dpg4x uses.

.TP
.B Individual Settings
.
If you want to set default settings for only one video (e.g. you may want to set
a specific thumbnail or subtitles file), you can select the media source and
press the "Set" button in the "Individual Settings" menu, at the bottom-right of
the MAIN section. A popup window will be opened, showing the same options
available in the default configuration sections, just perform the changes and
press save. The default settings will no longer be used for that media source,
only the settings you just set.

You can press "Set" again to perform another change or press "Delete" to delete
the individual configuration for that media source. If you are not sure about
if a media source will use individual or default settings you can check it in
the media sources list, in the MAIN section (second column).

.TP
.B Default Subtitles
.
If you have problems when adding subtitles to your videos, just include a
subtitles file in the folder where the original video file is. The name of the
subtitles file must be the same as the video file, except the extension. To
check that everything is ok, just play the video file with mplayer (e.g. with
the "Preview" button). As the last step, don't forget to switch to the SUBTITLES
section and select "Auto" as the subtitles source.

.
.SH "OPTIONS"
.
.TP
.B The --dpg option
.
You can encode the video files into DPG format with the --dpg option. For example:
.br
.I dpg4x --dpg input_file.avi 

This will create an DPG file called input_file.dpg or if that exists input_file~1.dpg.

The --dpg option will consider the options that have been set from the GUI, and either
use the default ones or, if defined, the individual settings for each input file.

.TP
.B The --avi option
.
You can encode the DPG video files to AVI format with the --avi option. For example:
.br
.I dpg4x --avi input_file.dpg 

This will create an AVI file called input_file.avi or if that exists input_file~1.avi.

The --avi option does not re-encode the video or the audio streams in the DPG
file. It only performs changes in the container, so all the properties like
framerate, video size, bitrates, quality, etc. will be the same.

.TP
.B The --thumbnail option
.
The --thumbnail option can replace the thumbnail embedded in a DPG file. Note
that files with DPG version 3 or less, cannot store a thumbnail.

To use the script, you need to set the image file. For
example:
.br
.I dpg4x --thumbnail=image_file.png file.dpg 

Most common image formats are supported: png, jpeg, gif, bmp, etc.

The GUI is started if no options are given.
.
.SH "FILES"
.
When you use dpg4x, it stores the current configuration in your home folder. The
full path of the configuration directory is:
.br
.I ~/.config/dpg4x/

The next time dpg4x is opened, it will restore the last saved configuration. You
can delete the configuration files to reset the default or the individual media
settings. But you should not edit the configuration file by hand. The default
settings are stored in a file named config.ini

You can also reset all the configuration to the default values by pressing the
"Reset to default" button in the MISC section.

The --avi option does not read the dpg4x configuration files. It stores the temporary
files in the output directory.
.
.SH "BUGS"
.
If you find a bug in dpg4x, please report it via the tracker:
.br
.UR http://sourceforge.net/projects/dpg4x/support
http://sourceforge.net/projects/dpg4x/support

Or email to a maintainer in order to get support. See the CREDITS file for
details. 
.
.SH "AUTHORS"
.
The credits file for DPG for X is available here:
.br
.I /usr/share/doc/dpg4x/CREDITS
.
.SH "SEE ALSO"
.
.UR http://sourceforge.net/projects/dpg4x/
http://sourceforge.net/projects/dpg4x/
.P
.


ChangeLog

                        #
 #####   #####    ####  #    #   #    #
 #    #  #    #  #    # #    #    #  #
 #    #  #    #  #      #######    ##
 #    #  #####   #  ###      #     ##
 #    #  #       #    #      #    #  #
 #####   #        ####       #   #    #


2010-02-05 dpg4x 1.0

    * First public release.

2010-05-29 dpg4x 1.1

    * You can now use Dvd and Vcd as media sources. 
    * The preview button now plays a real preview of the encoded file.
      You can setup the preview's size from the misc settings.
      The old behaviour is still available via the play button.
    * Fixed some bugs in the localization and subtitle's code.
    * Track values adjusted based on mplayer's documentation.
    * Added a new track setup for video.
    * DPG version 0 now only supports mono audio.

2010-06-20 dpg4x 1.1.1

    * Fixed bug 3011287 - GUI does not scale.

2010-08-25 dpg4x 1.2

    * The new dpg2avi script can convert DPG files to AVI format.
    * Added a "Media Info" button to gather information about sources.
    * Experimental OGG and GSM audio support. Disabled by default.
    * Various bug fixes.

2010-11-12 dpg4x 2.0

    * DPG for X can now be executed in Windows and MacOSX.
    * Support for individual media settings.
    * New language available: Catalan.
    * Lots and lots of other minor enhancements and bug fixes.

2011-04-10 dpg4x 2.1

    * New language available: Swedish.
    * Windows support heavily improved.
    * Project sources adapted for an easier package creation.
    * deb package available.
    * rpm package available.
    * Installer for Windows based on NSIS available.
    * New button to reset all configurations to the default values.
    * Better integration with some versions of mencoder (like Fedora's).
    * Lots of bug fixes.

2011-09-06 dpg4x 2.2

    * The new dpgimginjector script can replace the thumbnail image 
      embedded in a DPG file.
    * Thumbnail support improved. Optional antialiasing with PIL.
    * The Windows installer can handle updates.
    * Fixed bug 3306088.



dpg4x/Dpg4x.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#Boa:App:BoaApp

#----------------------------------------------------------------------------
# Name:         Dpg4x.py
# Purpose:      A dpg encoder for Linux (and maybe others).
#
# Author:       Félix Medrano Sanz
#
# Created:      
# RCS-ID:       $Id: Dpg4x.py $
# Copyright:    (c) 2009 Félix Medrano Sanz
# Licence:      GPL v3
#----------------------------------------------------------------------------

import sys
import os

# On Windows wxPython 2.9.1.1 works better than 2.8.11.0.
# In particuliar OutputTextDialog() is unreadable and unusable. You can only
# see the first couple of letters of the text and the close button is missing.

# wxPython running from py2exe fails with wxversion.select()
# See: http://www.wxpython.org/docs/api/wxversion-module.html
if not hasattr(sys, 'frozen'):
    import wxversion
    wxversion.select(['2.8','2.9'])
import wx


import MainFrame
import Globals
import DpgThumbnail
import Encoder

modules ={u'AddDvdDialog': [0,
                   u'A dialog to add Dvd media sources.',
                   u'AddDvdDialog.py'],
 u'AddVcdDialog': [0,
                   u'A dialog to add Vcd media sources.',
                   u'AddVcdDialog.py'],
 u'AudioPanel': [0, u'Panel with audio options.', u'AudioPanel.py'],
 u'ConfigurationManager': [0,
                           u'Manages the configuration variables.',
                           u'ConfigurationManager.py'],
 u'CustomFontSelector': [0,
                         u'Dialog to select fonts (only faces).',
                         u'CustomFontSelector.py'],
 u'CustomProgressDialog': [0,
                           u'Dialog to show the progress of the encoding.',
                           u'CustomProgressDialog.py'],
 u'DpgHeader': [0, u'DPG header management.', u'DpgHeader.py'],
 u'DpgInfoDialog': [0,
                    u'Dialog with a thumbail image and TextCtrls to show DPG info.',
                    u'moreControls/DpgInfoDialog.py'],
 u'DpgThumbnail': [0, u'Class to handle DPG thumbnails.', u'DpgThumbnail.py'],
 u'Encoder': [0, u'Performs the encoding duties.', u'Encoder.py'],
 u'FilesPanel': [0, u'Panel with files to be encoded.', u'FilesPanel.py'],
 u'Globals': [0,
              u'Source file with global variables and functions.',
              u'Globals.py'],
 u'MainFrame': [1, u'Main frame of Application.', u'MainFrame.py'],
 u'MediaAudioPanel': [0,
                      u'Panel with per-media audio options.',
                      u'MediaAudioPanel.py'],
 u'MediaMainFrame': [0,
                     u'Frame with per-media settings.',
                     u'MediaMainFrame.py'],
 u'MediaOtherPanel': [0,
                      u'Panel with per-media aditional options.',
                      u'MediaOtherPanel.py'],
 u'MediaSubtitlesPanel': [0,
                          u'Panel with per-media subtitle options.',
                          u'MediaSubtitlesPanel.py'],
 u'MediaVideoPanel': [0,
                      u'Panel with per-media video options.',
                      u'MediaVideoPanel.py'],
 u'OtherPanel': [0, u'Panel with aditional options.', u'OtherPanel.py'],
 u'OutputTextDialog': [0,
                       u"Dialog with a TextCtrl to show program's output.",
                       u'moreControls/OutputTextDialog.py'],
 u'Previewer': [0, u'Allows advanced preview options.', u'Previewer.py'],
 u'SubtitlesPanel': [0, u'Panel with subtitle options.', u'SubtitlesPanel.py'],
 u'TreeCtrlComboPopup': [0,
                         u'Popup control containing a TreeCtrl.',
                         u'TreeCtrlComboPopup.py'],
 u'VideoPanel': [0, u'Panel with video options.', u'VideoPanel.py']}
    
def checkDependencies():
    "Check that the mandatory dependecies are present"
    # Mandatory programs
    mandatory = ['mplayer','mencoder']
    # Check that all of them are present
    for program in mandatory:
        if not Globals.which(program):
            message = _(u'%s not found in PATH. Please install it.') % program
            # Show an error in the console
            Globals.debug(_(u'ERROR') + ': ' + message)
            # Show a dialog to the user
            dialog = wx.MessageDialog(None, message, _(u'ERROR'), style=wx.OK|wx.ICON_ERROR)
            dialog.ShowModal()
            sys.exit(1)

def checkArgs():
    usage=_(u"""Usage: dpg4x [options] file1 file2... (starts GUI if no options)

Options:
  -h, --help        show this help message and exit
  --thumbnail=FILE  use image file as thumbnail in DPG files
  --dpg             convert files to DPG, using settings from GUI
  --avi             convert DPG files to AVI, no transformation
""")

    # Naive parser because optparse does not support i18n
    options_image = None
    options_avi = False
    options_dpg = False
    args = []
    for arg in sys.argv[1:]:
        if arg == '--dpg':
            options_dpg = True
        elif arg == '--avi':
            options_avi = True
        elif arg.startswith('--thumbnail='): 
            options_image = arg.split('=')[1]
        elif arg == '-h' or arg == '--help':
            Globals.debug(usage)
            return False
        elif arg.startswith('-'):
            Globals.debug(_(u'Non supported option: %s') % arg)
            Globals.debug(usage)
            return False                
        else:
            args += [arg]

    if len(args)>0:
        try:
            if options_image:
                for a in args:
                    a = Globals.Decode(a)
                    Globals.debug(a)
                    DpgThumbnail.DpgInject(a,options_image,a)
                return False
            elif options_avi:
                for a in args:
                    a = Globals.Decode(a)
                    Globals.debug(a)
                    Encoder.encode_Dpg2Avi(a)
                    return False
            elif options_dpg:
                for a in args:
                    a = Globals.Decode(a)
                    Globals.debug(a)
                    Encoder.encode_files([a])
                return False
        except Exception, e:
            Globals.debug(_(u'ERROR') + ': ' + unicode(e.args[0]))
            return False
    return True

# Main function
if __name__ == '__main__':
    Globals.SetupTranslation()
    checkDependencies()
    firstExec = checkArgs() 
    application = wx.App(redirect=False,clearSigInt=False)
    while firstExec or Globals.restart:
        # Reload the Globals module on restart
        if Globals.restart:
            reload(Globals)
        firstExec = False
        mainFrame = MainFrame.create(None, Globals.getIconDir())
        mainFrame.Show()
        Globals.mainPanel = mainFrame
        application.SetTopWindow(mainFrame)
        application.MainLoop()
    sys.exit(0)
